File: ServerProject/src/main/java/com/example/meditime/dto/AdherenceLogDTO.java
// Amy Wickham 121785021
// File: AdherenceLogDTO.java
// Description: Data Transfer Object (DTO) for representing adherence log information in a simplified, serializable form
//              for client-side or REST API use. Part of the MediTime medication management system.

package com.example.meditime.dto;

import lombok.Data;
import com.example.meditime.model.AdherenceLog;

/**
 * A DTO (Data Transfer Object) that simplifies the AdherenceLog entity for use in APIs and front-end communication.
 * Only essential information is included to reduce payload size and increase clarity.
 */
@Data // Lombok annotation to automatically generate getters, setters, toString, equals, and hashCode
public class AdherenceLogDTO {

    // Unique identifier of the adherence log entry
    private Long adherenceId;

    // Full name of the user (e.g. carer or manager) who created or is associated with the log
    private String userFullName;

    // Name of the medication associated with the adherence log
    private String clientMedicationName;

    // Timestamp when the adherence was checked or recorded (as a String for easy JSON serialization)
    private String checkedTime;

    // Adherence rate (e.g. percentage of medication taken on time)
    private Double adherenceRate;

    /**
     * Converts an AdherenceLog entity into an AdherenceLogDTO.
     * This helps decouple the internal data model from what is exposed to the client.
     *
     * @param log the AdherenceLog entity to convert
     * @return a populated AdherenceLogDTO, or null if the input is null
     */
    public static AdherenceLogDTO fromEntity(AdherenceLog log) {
        if (log == null) return null;

        AdherenceLogDTO dto = new AdherenceLogDTO();
        dto.setAdherenceId(log.getAdherenceId());
        dto.setUserFullName(log.getUser().getName());
        dto.setClientMedicationName(log.getClientMedication().getMedication().getName());
        dto.setCheckedTime(log.getCheckedTime().toString());
        dto.setAdherenceRate(log.getAdherenceRate());

        return dto;
    }
}

File: Mobile/app/src/main/java/com/example/medicationapp/model/dto/AdherenceLogDTO.kt
package com.example.medicationapp.model.dto

/**
 * Data Transfer Object (DTO) for representing adherence log details in a user-friendly format.
 *
 * Typically used for displaying or transmitting adherence data to the UI or across layers.
 *
 * Fields:
 * - adherenceId: Unique identifier for the adherence log entry.
 * - userFullName: Full name of the user who recorded the adherence.
 * - clientMedicationName: Name of the medication associated with the adherence.
 * - checkedTime: Timestamp (as String) when the adherence was recorded.
 * - adherenceRate: Adherence percentage (e.g., 1.0 = 100%, 0.75 = 75%).
 * - scheduledTime: Scheduled time for taking the medication (as String).
 * - actualTime: Actual time the medication was taken (as String).
 * - status: Status of adherence (e.g., "On Time", "Late", "Missed").
 * - notes: Additional notes or comments related to the log.
 */

data class AdherenceLogDTO(
    val adherenceId: Int,
    val userFullName: String,
    val clientMedicationName: String,
    val checkedTime: String,
    val adherenceRate: Double,
    val scheduledTime: String,
    val actualTime: String,
    val status: String,
    val notes: String
)